com.teamscale.polarion.plugin/src/com/teamscale/polarion/plugin/WorkItemUpdatesServlet.java

```diff
@@ -0,0 +1,769 @@
+    // if there isn't, add an entry to the map flipping the ids (reverse)
+    for (ILinkedWorkItemStruct linkStruct : links) {
+      List<LinkBundle> linkBundles = backwardLinksTobeAdded.get(linkStruct.getLinkedItem().getId());
+      LinkBundle reverse = null;
+      if (linkBundles == null) {
+        reverse = new LinkBundle();
+        reverse.setAdded(added);
+        reverse.setRevision(revision);
+        reverse.setLinkedWorkItem(new LinkedWorkItem(workItemId, linkStruct.getLinkRole().getId()));
+        List<LinkBundle> newLinkBundles = new ArrayList<LinkBundle>();
+        newLinkBundles.add(reverse);
+        backwardLinksTobeAdded.put(linkStruct.getLinkedItem().getId(), newLinkBundles);
+      } else {
+        if (!alreadyHasLinkBundle(linkBundles, linkStruct, revision, added)) {
+          reverse = new LinkBundle();
+          reverse.setAdded(added);
+          reverse.setRevision(revision);
+          reverse.setLinkedWorkItem(
+              new LinkedWorkItem(workItemId, linkStruct.getLinkRole().getId()));
+          linkBundles.add(reverse);
+        }
+      }
+    }
+  }
+
+  /**
+   * Helper method called by {@link #updateOppositeLinksMap()} to check if a LinkBundle was already
+   * created in the map *
+   */
+  private boolean alreadyHasLinkBundle(
+      List<LinkBundle> linkBundles,
+      ILinkedWorkItemStruct linkStruct,
+      String revision,
+      boolean added) {
+    if (linkBundles == null) return false;
+    for (LinkBundle linkBundle : linkBundles) {
+      if (linkBundle.getRevision().equals(revision)
+          && linkBundle.isAdded() == added
+          && linkBundle.getLinkedWorkItem().getLinkRoleId().equals(linkStruct.getLinkRole().getId())
+          && linkBundle.getLinkedWorkItem().getId().equals(linkStruct.getLinkedItem().getId())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Binary search to cut down the search space since the list is ordered in ascending order* */
+  private int searchIndexWorkItemHistory(IPObjectList<IWorkItem> workItemHistory) {
+
+    if (workItemHistory == null) return -1;
+
+    Integer lastUpdateInt = Integer.valueOf(lastUpdate);
+    Integer endRevisionInt = Integer.valueOf(endRevision);
+    Integer lastItemRevision =
+        Integer.valueOf(workItemHistory.get(workItemHistory.size() - 1).getRevision());
+    Integer firstItemVersion = Integer.valueOf(workItemHistory.get(0).getRevision());
+
+    // Short circuit: don't need to binarysearch if you're looking for all history
+    if (lastUpdateInt <= 0) return 0;
+
+    // Short circuit: don't need to search if lastUpdate points to the last version of the item
+    if (lastItemRevision == lastUpdateInt) return workItemHistory.size() - 1;
+
+    // Short circuit: don't search if all changes are before lastUpdate or after endRevision
+    if (lastItemRevision < lastUpdateInt || firstItemVersion > endRevisionInt) return -1;
+
+    int left = 0;
```

Comment:
Maybe use `Collections.binarySearch` here?

Reply:
It's not a plain binary search. There's a slight variation. The if inside the else-if and the two blocks after the while.